crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -165,12 +160,49 @@ impl AuthorityStorePruner {
         Ok(pruned)
     }
 
+    fn process_checkpoints(
+        perpetual_db: &Arc<AuthorityPerpetualTables>,
+        checkpoint_store: &Arc<CheckpointStore>,
+        num_epochs_to_retain: u64,
+    ) -> anyhow::Result<()> {
+        let mut pruned_seq_number = checkpoint_store
+            .get_highest_pruned_checkpoint_seq_number()?
+            .unwrap_or_default();
+        let current_epoch = checkpoint_store
+            .get_highest_executed_checkpoint()?
+            .map(|c| c.epoch())
+            .unwrap_or_default();
+        loop {
+            let Some(checkpoint) = checkpoint_store.get_checkpoint_by_sequence_number(pruned_seq_number + 1)?
+                else {return Ok(());};
```

Comment:
This approach certainly works. But how about instead of iterating checkpoints from last pruned seq num to highest epoch checkpoint possible, we iterate in the other direction? We are going to be adding a lot of tombstones in the current approach for the same object id in every checkpoint it was mutated in. Whereas iterating in the other direction will allow us to add just one tombstone per object id (latest modified seq to zero) by keeping track of touched object ids in some set data structures like hashset, bloom filter, bitmap, etc

Reply:
it makes sense!
We still want to try to switch from delete ranges to point-deletes though. In that case, there won't be much of a difference. Each delete will be a separate tombstone and then it's just easier to reason about this way.
But yeah, we may stuck with delete ranges, so let's definitely keep this optimization in mind